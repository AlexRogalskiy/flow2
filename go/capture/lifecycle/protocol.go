package lifecycle

import (
	"encoding/json"
	"fmt"

	pc "github.com/estuary/flow/go/protocols/capture"
	pf "github.com/estuary/flow/go/protocols/flow"
)

// StageCaptured potentially sends a previously staged Captured into the stream,
// and then stages its arguments into response.Captured.
func StageCaptured(
	stream interface {
		Send(*pc.CaptureResponse) error
	},
	response **pc.CaptureResponse,
	binding int,
	document json.RawMessage,
) error {
	// Send current |response| if we would re-allocate.
	if *response != nil {
		var rem int
		if l := (*response).Captured; int(l.Binding) != binding {
			rem = -1 // Must flush this response.
		} else if cap(l.DocsJson) != len(l.DocsJson) {
			rem = cap(l.Arena) - len(l.Arena)
		}
		if rem < len(document) {
			if err := stream.Send(*response); err != nil {
				return fmt.Errorf("sending Captured response: %w", err)
			}
			*response = nil
		}
	}

	if *response == nil {
		*response = &pc.CaptureResponse{
			Captured: &pc.CaptureResponse_Captured{
				Binding:  uint32(binding),
				Arena:    make(pf.Arena, 0, arenaSize),
				DocsJson: make([]pf.Slice, 0, sliceSize),
			},
		}
	}

	var l = (*response).Captured
	l.DocsJson = append(l.DocsJson, l.Arena.Add(document))

	return nil
}

// WriteCommit flushes a pending Captured response, and sends a Commit response
// with the provided driver checkpoint.
func WriteCommit(
	stream interface {
		Send(*pc.CaptureResponse) error
	},
	response **pc.CaptureResponse,
	commit *pc.CaptureResponse_Commit,
) error {
	// Flush partial Captured response, if required.
	if *response != nil {
		if err := stream.Send(*response); err != nil {
			return fmt.Errorf("flushing final Captured response: %w", err)
		}
		*response = nil
	}

	if err := stream.Send(&pc.CaptureResponse{
		Commit: commit,
	}); err != nil {
		return fmt.Errorf("sending Commit response: %w", err)
	}

	return nil
}

const (
	arenaSize = 16 * 1024
	sliceSize = 32
)
