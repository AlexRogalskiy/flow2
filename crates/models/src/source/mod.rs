use crate::tables;

mod loader;
pub mod scenarios;
mod scope;
pub mod specs;
pub mod wrappers;

pub use loader::{FetchResult, LoadError, Loader};
pub use scope::Scope;

pub use specs::*;
pub use wrappers::*;

#[derive(Default, Debug)]
pub struct Tables {
    pub captures: tables::Captures,
    pub collections: tables::Collections,
    pub derivations: tables::Derivations,
    pub endpoints: tables::Endpoints,
    pub errors: tables::Errors,
    pub fetches: tables::Fetches,
    pub imports: tables::Imports,
    pub materializations: tables::Materializations,
    pub nodejs_dependencies: tables::NodeJSDependencies,
    pub projections: tables::Projections,
    pub resources: tables::Resources,
    pub schema_docs: tables::SchemaDocs,
    pub test_steps: tables::TestSteps,
    pub transforms: tables::Transforms,
}

impl Tables {
    // Access Tables as an array of dynamic TableObj instances.
    pub fn as_tables(&self) -> Vec<&dyn tables::TableObj> {
        // This de-structure ensures we can't fail to update if fields change.
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }

    // Access Tables as an array of mutable dynamic TableObj instances.
    pub fn as_tables_mut(&mut self) -> Vec<&mut dyn tables::TableObj> {
        let Tables {
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        } = self;

        vec![
            captures,
            collections,
            derivations,
            endpoints,
            errors,
            fetches,
            imports,
            materializations,
            nodejs_dependencies,
            projections,
            resources,
            schema_docs,
            test_steps,
            transforms,
        ]
    }
}
