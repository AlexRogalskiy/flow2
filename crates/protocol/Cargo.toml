[package]
name = "protocol"
version = "0.0.0"
authors = ["Estuary Technologies, Inc"]
edition = "2018"

[dependencies]
prost = "*"
prost-types = "*"
serde = "*"
tonic = "*"

[dev-dependencies]
insta = "*"
serde_json = "*"

[build-dependencies]
tonic-build = "*"
