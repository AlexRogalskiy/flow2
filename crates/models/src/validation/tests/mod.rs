use crate::source;
use crate::tables;
use crate::validation;

use futures::{future::LocalBoxFuture, FutureExt};
use lazy_static::lazy_static;
use protocol::materialize;
use serde_json::Value;
use std::collections::HashMap;

lazy_static! {
    static ref GOLDEN: Value = serde_yaml::from_slice(include_bytes!("model.yaml")).unwrap();
}

#[test]
fn test_golden_all_visits() {
    let (source_tables, built_tables) = run_test(GOLDEN.clone());
    insta::assert_debug_snapshot!(built_tables);
    insta::assert_debug_snapshot!(source_tables);
}

#[test]
fn test_database_round_trip() {
    let (source_tables, built_tables) = run_test(GOLDEN.clone());

    // Round-trip source and built tables through the database, verifying equality.
    let db = rusqlite::Connection::open(":memory:").unwrap();
    tables::persist_tables(&db, &source_tables.as_tables()).unwrap();
    let mut restored_tables = source::Tables::default();
    tables::load_tables(&db, restored_tables.as_tables_mut().as_mut_slice()).unwrap();

    assert_eq!(
        format!("{:#?}", source_tables),
        format!("{:#?}", restored_tables)
    );

    let db = rusqlite::Connection::open(":memory:").unwrap();
    tables::persist_tables(&db, &built_tables.as_tables()).unwrap();
    let mut restored_tables = validation::Tables::default();
    tables::load_tables(&db, restored_tables.as_tables_mut().as_mut_slice()).unwrap();

    assert_eq!(
        format!("{:#?}", built_tables),
        format!("{:#?}", restored_tables)
    );
}

#[test]
fn test_golden_error() {
    run_test_errors(&GOLDEN, "{}");
}

#[test]
fn test_invalid_collection_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  collections:
    good: &spec
      schema: test://int-string.schema
      key: [/int]
      store: { name: storeEndpoint }

    "": *spec
    bad name: *spec
    bad!name: *spec

    # We require a sequence of non-empty tokens, separated by exactly one '/'.
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string & others.
    testing: *spec

    # Illegal duplicates under naming collation.
    testing/int-sTRinG: *spec
    testing/Int-Halve: *spec
"#,
    );
}

#[test]
fn test_invalid_partition_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      projections:
        good: &spec
          location: /bit
          partition: true

        "": *spec
        inv alid: *spec
        inv!alid: *spec
        inv/alid: *spec

        # Illegal duplicates under collation.
        INT: /int
        bIt: /bit

        # Attempt to re-map a canonical projection.
        str: /int
"#,
    );
}

#[test]
fn test_invalid_transform_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          good: &spec
            source:
              name: testing/int-string
            publish:
              nodeJS: "something something"

          "": *spec
          inv alid: *spec
          inv!alid: *spec
          inv/alid: *spec

          # Illegal duplicate under collation.
          reVeRsEIntString: *spec
"#,
    );
}

#[test]
fn test_invalid_endpoint_names_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  import:
    - test://more-endpoints
test://more-endpoints:
  endpoints:
    good: &spec
      s3:
        bucket: a-bucket
        prefix: and-prefix

    "": *spec
    inv alid: *spec
    inv!alid: *spec
    inv/alid: *spec

    # Illegal duplicates under collation.
    StoreEndpoinT: *spec
    CAPtUReEndpoINT: *spec
"#,
    );
}

#[test]
fn test_invalid_capture_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  captures:
    good: &spec
      target:
        name: testing/int-string
      pushAPI: {}

    "": *spec
    bad name: *spec
    bad!name: *spec
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string/pull & push.
    testing: *spec

    # Illegal duplicates under naming collation.
    testing/int-sTRinG/pUll: *spec
    testing/Int-strINg/PuSH: *spec
"#,
    );
}

#[test]
fn test_invalid_materialization_names_prefixes_and_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  materializations:
    good: &spec
      source:
        name: testing/int-string
      endpoint:
        name: materializeEndpoint
        config: { fixture: one }

    "": *spec
    bad name: *spec
    bad!name: *spec
    bad//name: *spec
    bad/name/: *spec
    /bad/name: *spec

    # Invalid prefix of testing/int-string.
    testing: *spec

    # Illegal duplicate under naming collation.
    testing/int-sTRinG: *spec
"#,
    );
}

#[test]
fn test_transform_source_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source: { name: testinG/Int-String }
          halveSelf:
            source: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_capture_target_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  captures:
    testing/int-string/pull:
      target: { name: testiNg/int-strinK }
    testing/int-string/push:
      target: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_capture_endpoint_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  captures:
    testing/int-string/pull:
      endpoint: { name: CaptureEndpoit }
    testing/int-string/push:
      endpoint: { name: wildlyOffName }
      pushAPI: null
"#,
    );
}

#[test]
fn test_capture_endpoint_wrong_type() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  endpoints:
    captureEndpoint:
      s3: null
      postgres:
        host: a-host
        user: a-user
        password: a-password
"#,
    );
}

#[test]
fn test_capture_target_is_derivation_and_missing_imports() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-capture:
  import: null
  captures:
    testing/int-string/pull:
      target: { name: testing/int-reverse }
"#,
    );
}

#[test]
fn test_capture_duplicates() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  captures:
    testing/int-string/another/pull:
      target:
        name: testing/int-string
      endpoint: { name: captureEndpoint }
"#,
    );
}

#[test]
fn test_collection_store_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      store: { name: sToreEndpoin }

test://int-reverse:
  collections:
    testing/int-reverse:
      store: { name: WildlyOffName }
"#,
    );
}

#[test]
fn test_use_without_import() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  import: [] # Clear.
  collections:
    testing/int-string:
      store: { name: s3WithoutImport }

test://int-reverse:
  import: [] # Clear.
  endpoints:
    s3WithoutImport:
      s3:
        bucket: a-bucket
        prefix: and-prefix
"#,
    );
}

#[test]
fn test_schema_fragment_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      schema: test://int-string.schema#/not/found

test://int-string-materialization:
  materializations:
    testing/int-string: null # Omit downstream errors.

test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source:
              schema: test://int-string-len.schema#/not/found
"#,
    );
}

#[test]
fn test_keyed_location_wrong_type() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string.schema:
  properties:
    int: { type: [number, object] }
"#,
    );
}

#[test]
fn test_unknown_locations() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string:
  collections:
    testing/int-string:
      key: [/int, /unknown/key]
      projections:
        Unknown: /unknown/projection

test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/len, /int, /unknown/shuffle]
"#,
    );
}

#[test]
fn test_shuffle_key_length_mismatch() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/len]
          halveSelf:
            shuffle:
              key: [/len, /int]
"#,
    );
}

#[test]
fn test_shuffle_key_types_mismatch() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            shuffle:
              key: [/int, /str]
          halveSelf:
            shuffle:
              key: [/str, /int]
"#,
    );
}

#[test]
fn test_shuffle_key_empty() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            shuffle: {key: []}
"#,
    );
}

#[test]
fn test_partition_selections() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        transform:
          halveIntString:
            source:
              partitions:
                include:
                  bit: [true, 42, ""]
                  Int: [15, true]
                  Unknown: ["whoops"]
                exclude:
                  bit: [false, "a string"]
                  Int: [false, "", 16]
                  AlsoUnknown: ["whoops"]
"#,
    );
}

#[test]
fn test_redundant_source_schema_and_shuffle() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            source:
              name: testing/int-string
              schema: test://int-string.schema
            shuffle:
              key: [/int]
"#,
    );
}

#[test]
fn test_incompatible_collection_store() {
    run_test_errors(
        &GOLDEN,
        r#"
test://root:
  endpoints:
    storeEndpoint:
      s3: null
      postgres:
        host: a-host
        user: a-user
        password: a-password
"#,
    );
}

#[test]
fn test_must_have_update_or_publish() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-reverse:
  collections:
    testing/int-reverse:
      derivation:
        transform:
          reverseIntString:
            publish: null
            update: null
"#,
    );
}

#[test]
fn test_invalid_initial_register() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve:
  collections:
    testing/int-halve:
      derivation:
        register:
          initial: "should be an integer"
"#,
    );
}

#[test]
fn test_shape_inspections() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-len.schema:
  properties:
    int:
      reduce: { strategy: set }
    str:
      reduce: { strategy: sum }
"#,
    );
}

#[test]
fn test_schema_reference_verification() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-len.schema:
  $ref: test://int-string.schema#/whoops
"#,
    );
}

#[test]
fn test_materialization_source_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      source: { name: testiNg/int-strinK }

test://int-halve-materialization:
  materializations:
    testing/int-halve:
      source: { name: wildly/off/name }
"#,
    );
}

#[test]
fn test_materialization_endpoint_not_found() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      endpoint: { name: MaterializeEndpoit }

test://int-halve-materialization:
  materializations:
    testing/int-halve:
      endpoint: { name: wildlyOffName }
"#,
    );
}

#[test]
fn test_materialization_field_errors() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-halve-materialization:
  materializations:
    testing/int-halve:
      fields:
        include:
          int: {} # Include and exclude.
          biT: {} # Unknown.
        exclude:
          - BiTT # Unknown.
          - WildlyOffName # Also unknown.
          - int
"#,
    );
}

#[test]
fn test_materialization_driver_returns_error() {
    run_test_errors(
        &GOLDEN,
        r#"
driver:
  materializations:
    one:
      constraints: null
      error: "A driver error!"
"#,
    );
}

#[test]
fn test_materialization_driver_unknown_constraint() {
    run_test_errors(
        &GOLDEN,
        r#"
driver:
  materializations:
    one:
      constraints:
        str: { type: 99, reason: "whoops" }
"#,
    );
}

#[test]
fn test_materialization_driver_conflicts() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-materialization:
  materializations:
    testing/int-string:
      fields:
        include:
          str: {}
        exclude:
          - bit
          - Int
driver:
  materializations:
    one:
      constraints:
        flow_document: { type: 1, reason: "location required" }
        Int: { type: 1, reason: "location required" }

        int: { type: 5, reason: "field unsatisfiable" }
        str: { type: 4, reason: "field forbidden" }
        bit: { type: 0, reason: "field required" }
        Unknown: { type: 0, reason: "whoops" }
"#,
    );
}

#[test]
fn test_test_step_unknown_collection() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - ingest:
          collection: testinG/Int-strin
          documents: []
      - verify:
          collection: wildly/Off/Name
          documents: []
"#,
    );
}

#[test]
fn test_test_step_ingest_schema_error() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - ingest:
          collection: testing/int-string
          documents:
            - {int: 42, str_whoops: "string A", bit: true}
            - {int: 52, str_whoops: "string B", bit: true}
"#,
    );
}

#[test]
fn test_test_step_verify_key_order() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - verify:
          collection: testing/int-string
          documents: [{int: 52}, {int: 62}, {int: 42}]
"#,
    );
}

#[test]
fn test_test_step_verify_selector() {
    run_test_errors(
        &GOLDEN,
        r#"
test://int-string-tests:
  tests:
    "A Test":
      - verify:
          collection: testing/int-string
          documents: []
          partitions:
            include:
              bit: [true, 42, ""]
              Int: [15, true]
              Unknown: ["whoops"]
            exclude:
              bit: [false, "a string"]
              Int: [false, "", 16]
              AlsoUnknown: ["whoops"]
"#,
    );
}

#[derive(serde::Deserialize)]
#[serde(rename_all = "camelCase")]
struct MockDriverCalls {
    materializations: HashMap<String, MockMaterializationValidateCall>,
}

#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
struct MockMaterializationValidateCall {
    endpoint: source::EndpointType,
    config: serde_json::Value,
    #[serde(default)]
    constraints: HashMap<String, materialize::Constraint>,
    #[serde(default)]
    error: Option<String>,
}

impl validation::Drivers for MockDriverCalls {
    fn validate_materialization(
        &self,
        endpoint_type: source::EndpointType,
        endpoint_config: serde_json::Value,
        _request: materialize::ValidateRequest,
    ) -> LocalBoxFuture<Result<materialize::ValidateResponse, validation::BoxError>> {
        async move {
            for (_key, call) in &self.materializations {
                if (call.endpoint, &call.config) != (endpoint_type, &endpoint_config) {
                    continue;
                }

                if let Some(err) = &call.error {
                    return Err(err.as_str().into());
                } else {
                    return Ok(materialize::ValidateResponse {
                        constraints: call.constraints.clone(),
                    });
                }
            }
            return Err("driver fixture not found".into());
        }
        .boxed_local()
    }
}

fn run_test(mut fixture: Value) -> (source::Tables, validation::Tables) {
    // Extract out driver mock call fixtures.
    let mock_calls = fixture
        .get_mut("driver")
        .map(|d| d.take())
        .unwrap_or_default();
    let mock_calls: MockDriverCalls = serde_json::from_value(mock_calls).unwrap();

    let source_tables = source::scenarios::evaluate_fixtures(Default::default(), &fixture);
    if !source_tables.errors.is_empty() {
        eprint!("{:?}", &source_tables);
        panic!("unexpected fixture load error");
    }
    let built_tables = validation::validate(&mock_calls, &source_tables);

    (source_tables, built_tables)
}

fn run_test_errors(fixture: &Value, patch: &str) {
    let mut fixture = fixture.clone();
    let patch: Value = serde_yaml::from_str(patch).unwrap();
    json_patch::merge(&mut fixture, &patch);

    let (_, built_tables) = run_test(fixture);
    insta::assert_debug_snapshot!(built_tables.errors);
}
