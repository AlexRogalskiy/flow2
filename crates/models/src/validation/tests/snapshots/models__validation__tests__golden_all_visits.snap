---
source: crates/models/src/validation/tests/mod.rs
expression: tables
---
Tables {
    built_collections: [
        BuiltCollection {
            scope: test://int-reverse#/collections/testing~1int-reverse,
            collection: testing/int-reverse,
            spec: {"name":"testing/int-reverse","schema_uri":"test://int-string.schema","key_ptrs":["/int"],"uuid_ptr":"/_meta/uuid","projections":[{"ptr":"/bit","field":"bit","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["boolean"],"must_exist":true}},{"ptr":"","field":"flow_document","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["object"],"must_exist":true}},{"ptr":"/int","field":"int","user_provided":false,"is_partition_key":false,"is_primary_key":true,"inference":{"types":["integer"],"must_exist":true}},{"ptr":"/str","field":"str","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["string"],"must_exist":true,"string":{"is_base64":false}}}],"journal_spec":{"name":"","replication":3,"labels":null,"fragment":{"length":536870912,"compression_codec":5,"stores":["s3://a-bucket/and-prefix"],"refresh_interval":"300s","retention":null,"flush_interval":"3600s","path_postfix_template":"utc_date={{.Spool.FirstAppendTime.Format \"2006-01-02\"}}/utc_hour={{.Spool.FirstAppendTime.Format \"15\"}}"},"flags":0,"max_append_rate":0},"ack_json_template":"{\"_meta\":{\"ack\":true,\"uuid\":\"DocUUIDPlaceholder-329Bb50aa48EAa9ef\"}}"},
        },
        BuiltCollection {
            scope: test://int-string#/collections/testing~1int-string,
            collection: testing/int-string,
            spec: {"name":"testing/int-string","schema_uri":"test://int-string.schema","key_ptrs":["/int"],"uuid_ptr":"/_meta/uuid","partition_fields":["bit"],"projections":[{"ptr":"/int","field":"Int","user_provided":true,"is_partition_key":false,"is_primary_key":true,"inference":{"types":["integer"],"must_exist":true}},{"ptr":"/bit","field":"bit","user_provided":true,"is_partition_key":true,"is_primary_key":false,"inference":{"types":["boolean"],"must_exist":true}},{"ptr":"","field":"flow_document","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["object"],"must_exist":true}},{"ptr":"/int","field":"int","user_provided":false,"is_partition_key":false,"is_primary_key":true,"inference":{"types":["integer"],"must_exist":true}},{"ptr":"/str","field":"str","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["string"],"must_exist":true,"string":{"is_base64":false}}}],"journal_spec":{"name":"","replication":3,"labels":null,"fragment":{"length":536870912,"compression_codec":5,"stores":["s3://a-bucket/and-prefix"],"refresh_interval":"300s","retention":null,"flush_interval":"3600s","path_postfix_template":"utc_date={{.Spool.FirstAppendTime.Format \"2006-01-02\"}}/utc_hour={{.Spool.FirstAppendTime.Format \"15\"}}"},"flags":0,"max_append_rate":0},"ack_json_template":"{\"_meta\":{\"ack\":true,\"uuid\":\"DocUUIDPlaceholder-329Bb50aa48EAa9ef\"}}"},
        },
        BuiltCollection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            spec: {"name":"testing/int-halve","schema_uri":"test://int-string-len.schema","key_ptrs":["/int"],"uuid_ptr":"/_meta/uuid","partition_fields":["Len"],"projections":[{"ptr":"/extra","field":"Extra","user_provided":true,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["boolean"],"must_exist":false}},{"ptr":"/len","field":"Len","user_provided":true,"is_partition_key":true,"is_primary_key":false,"inference":{"types":["integer"],"must_exist":true}},{"ptr":"","field":"Root","user_provided":true,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["object"],"must_exist":true}},{"ptr":"/bit","field":"bit","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["boolean"],"must_exist":true}},{"ptr":"/extra","field":"extra","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["boolean"],"must_exist":false}},{"ptr":"","field":"flow_document","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["object"],"must_exist":true}},{"ptr":"/int","field":"int","user_provided":false,"is_partition_key":false,"is_primary_key":true,"inference":{"types":["integer"],"must_exist":true}},{"ptr":"/len","field":"len","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["integer"],"must_exist":true}},{"ptr":"/str","field":"str","user_provided":false,"is_partition_key":false,"is_primary_key":false,"inference":{"types":["string"],"must_exist":true,"string":{"is_base64":false}}}],"journal_spec":{"name":"","replication":3,"labels":null,"fragment":{"length":536870912,"compression_codec":5,"stores":["s3://a-bucket/and-prefix"],"refresh_interval":"300s","retention":null,"flush_interval":"3600s","path_postfix_template":"utc_date={{.Spool.FirstAppendTime.Format \"2006-01-02\"}}/utc_hour={{.Spool.FirstAppendTime.Format \"15\"}}"},"flags":0,"max_append_rate":0},"ack_json_template":"{\"_meta\":{\"ack\":true,\"uuid\":\"DocUUIDPlaceholder-329Bb50aa48EAa9ef\"}}"},
        },
    ],
    built_materializations: [
        BuiltMaterialization {
            scope: test://int-string-materialization#/materializations/testing~1int-string,
            materialization: testing/int-string,
            collection: testing/int-string,
            endpoint_config: {"bucket":"a-bucket","fixture":"one","prefix":"and-prefix"},
            field_selection: {"keys":["int"],"values":["str"],"document":"flow_document","field_config":{"str":"{\"pass\":\"through\"}"}},
        },
        BuiltMaterialization {
            scope: test://int-halve-materialization#/materializations/testing~1int-halve,
            materialization: testing/int-halve,
            collection: testing/int-halve,
            endpoint_config: {"bucket":"a-bucket","fixture":"two","prefix":"and-prefix"},
            field_selection: {"keys":["int"],"values":["Len","Extra"],"document":"Root","field_config":{"Len":"{}"}},
        },
    ],
    errors: [],
    implicit_projections: [
        Projection {
            scope: test://int-reverse#/collections/testing~1int-reverse,
            collection: testing/int-reverse,
            field: bit,
            location: /bit,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-reverse#/collections/testing~1int-reverse,
            collection: testing/int-reverse,
            field: flow_document,
            location: ,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-reverse#/collections/testing~1int-reverse,
            collection: testing/int-reverse,
            field: int,
            location: /int,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-reverse#/collections/testing~1int-reverse,
            collection: testing/int-reverse,
            field: str,
            location: /str,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-string#/collections/testing~1int-string,
            collection: testing/int-string,
            field: flow_document,
            location: ,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-string#/collections/testing~1int-string,
            collection: testing/int-string,
            field: int,
            location: /int,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-string#/collections/testing~1int-string,
            collection: testing/int-string,
            field: str,
            location: /str,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            field: bit,
            location: /bit,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            field: extra,
            location: /extra,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            field: flow_document,
            location: ,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            field: int,
            location: /int,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            field: len,
            location: /len,
            partition: 0,
            user_provided: 0,
        },
        Projection {
            scope: test://int-halve#/collections/testing~1int-halve,
            collection: testing/int-halve,
            field: str,
            location: /str,
            partition: 0,
            user_provided: 0,
        },
    ],
    inferences: [
        Inference {
            schema: test://int-halve?ptr=/collections/testing~1int-halve/derivation/register/schema,
            location: ,
            spec: {"types":["integer"],"must_exist":true},
        },
        Inference {
            schema: test://int-reverse?ptr=/collections/testing~1int-reverse/derivation/register/schema,
            location: ,
            spec: {"types":["array","boolean","null","number","object","string"],"must_exist":true,"string":{"is_base64":false}},
        },
        Inference {
            schema: test://int-string-len.schema,
            location: /bit,
            spec: {"types":["boolean"],"must_exist":true},
        },
        Inference {
            schema: test://int-string-len.schema,
            location: /extra,
            spec: {"types":["boolean"],"must_exist":false},
        },
        Inference {
            schema: test://int-string-len.schema,
            location: ,
            spec: {"types":["object"],"must_exist":true},
        },
        Inference {
            schema: test://int-string-len.schema,
            location: /int,
            spec: {"types":["integer"],"must_exist":true},
        },
        Inference {
            schema: test://int-string-len.schema,
            location: /len,
            spec: {"types":["integer"],"must_exist":true},
        },
        Inference {
            schema: test://int-string-len.schema,
            location: /str,
            spec: {"types":["string"],"must_exist":true,"string":{"is_base64":false}},
        },
        Inference {
            schema: test://int-string.schema,
            location: /bit,
            spec: {"types":["boolean"],"must_exist":true},
        },
        Inference {
            schema: test://int-string.schema,
            location: ,
            spec: {"types":["object"],"must_exist":true},
        },
        Inference {
            schema: test://int-string.schema,
            location: /int,
            spec: {"types":["integer"],"must_exist":true},
        },
        Inference {
            schema: test://int-string.schema,
            location: /str,
            spec: {"types":["string"],"must_exist":true,"string":{"is_base64":false}},
        },
    ],
}
